cylc/flow/network/scan.py

```diff
@@ -13,302 +13,361 @@
-            I.E. an async function.
-        iterator (iter):
-            Should yield tuples to be passed into the coroutine.
+        listing (list):
+            A listing of the directory in question as a list of
+            ``pathlib.Path`` objects.
+
+    Returns:
+        bool - True if the listing indicates that this is a flow directory.
+
+    """
+    listing = {
+        path.name
+        for path in listing
+    }
+    return (
+        SERVICE.name in listing
+        or SUITERC.name in listing  # cylc7 flow definition file name
+        or 'flow.cylc' in listing  # cylc8 flow definition file name
+        or 'log' in listing
+    )
+
+
+@pipe
+async def scan(run_dir=None, scan_dir=None, max_depth=4):
+    """List flows installed on the filesystem.
+
+    Args:
+        run_dir (pathlib.Path):
+            The run dir to look for workflows in, defaults to ~/cylc-run.
+
+            All workflow registrations will be given relative to this path.
+        scan_dir(pathlib.Path):
+            The directory to scan for workflows in.
+
+            Use in combination with run_dir if you want to scan a subdir
+            within the run_dir.
+        max_depth (int):
+            The maximum number of levels to descend before bailing.
+
+            * ``max_depth=1`` will pick up top-level suites (e.g. ``foo``).
+            * ``max_depth=2`` will pick up nested suites (e.g. ``foo/bar``).
 
     Yields:
-        list - List of results.
+        dict - Dictionary containing information about the flow.
+
+    """
+    if not run_dir:
+        run_dir = Path(
+            platform_from_name()['run directory'].replace('$HOME', '~')
+        ).expanduser()
+    if not scan_dir:
+        scan_dir = run_dir
+    stack = asyncio.Queue()
+    for subdir in await scandir(scan_dir):
+        if subdir.is_dir():
+            await stack.put((1, subdir))
+
+    # for path in stack:
+    async for depth, path in asyncqgen(stack):
+        contents = await scandir(path)
+        if dir_is_flow(contents):
+            # this is a flow directory
+            yield {
+                'name': str(path.relative_to(run_dir)),
+                'path': path,
+            }
+        elif depth < max_depth:
+            # we may have a nested flow, lets see...
+            for subdir in contents:
+                if subdir.is_dir():
+                    await stack.put((depth + 1, subdir))
+
+
+def join_regexes(*patterns):
+    """Combine multiple regexes using OR logic."""
+    return (re.compile(rf'({"|".join(patterns)})'),), {}
+
+
+@pipe(preproc=join_regexes)
+async def filter_name(flow, pattern):
+    """Filter flows by name.
 
-    Example:
-        >>> async def square(number): return number ** 2
-        >>> generator = async_map(square, ((i,) for i in range(5)))
-        >>> list(generator)
-        [0, 1, 4, 9, 16]
+    Args:
+        flow (dict):
+            Flow information dictionary, provided by scan through the pipe.
+        pattern (str):
```

Comment:
s/`str`/`re.Pattern`? IDE complaining about calling `.match` on a `str` type

Reply:
Good spotting.

Reply:
Oh, actually, I'm not sure about that.

The function accepts `re.Pattern` objects, however, the user passes `str` objects because the `pipe` decorator pre-processes the arguments.